common/metrics/metrics.go

```diff
@@ -102,3 +138,188 @@ func getMigrationDataSourceDetails(driver string, migrationData *migration.Migra
 		return migration.MigrationData_SOURCE_CONNECTION_MECHANISM_UNSPECIFIED.Enum(), migration.MigrationData_SOURCE_UNSPECIFIED.Enum()
 	}
 }
+
+func getDashboardClient(ctx context.Context) *dashboard.DashboardsClient {
+	if dashboardClient == nil {
+		once.Do(func() {
+			dashboardClient, _ = dashboard.NewDashboardsClient(ctx)
+		})
+		return dashboardClient
+	}
+	return dashboardClient
+}
+
+// CreateDataflowShardMonitoringDashboard returns a monitoring dashboard for a single shard
+func (resourceIds MonitoringMetricsResources) CreateDataflowShardMonitoringDashboard(ctx context.Context) (*dashboardpb.Dashboard, error) {
+	var mosaicLayoutTiles []*dashboardpb.MosaicLayout_Tile
+	var heightOffset int32 = 0
+	// create independent metrics tiles
+	independentMetricsTiles := createShardIndependentMetrics(resourceIds)
+	heightOffset += setWidgetPositions(independentMetricsTiles, heightOffset)
+	mosaicLayoutTiles = append(mosaicLayoutTiles, independentMetricsTiles...)
+	// create dataflow metrics
+	dataflowMetricsTiles := createShardDataflowMetrics(resourceIds)
+	dataflowMetricsGroupTile, heightOffset := createCollapsibleGroupTile(TileInfo{Title: fmt.Sprintf("Dataflow Job: %s", resourceIds.DataflowJobId)}, dataflowMetricsTiles, heightOffset)
+	mosaicLayoutTiles = append(append(mosaicLayoutTiles, dataflowMetricsTiles...), dataflowMetricsGroupTile)
+	// create datastream metrics tiles
+	datastreamMetricsTiles := createShardDatastreamMetrics(resourceIds)
+	datastreamMetricsGroupTile, heightOffset := createCollapsibleGroupTile(TileInfo{Title: fmt.Sprintf("Datastream: %s", resourceIds.DatastreamId)}, datastreamMetricsTiles, heightOffset)
+	mosaicLayoutTiles = append(append(mosaicLayoutTiles, datastreamMetricsTiles...), datastreamMetricsGroupTile)
+	// create gcs bucket metrics tiles
+	gcsMetricsTiles := createShardGcsMetrics(resourceIds)
+	gcsMetricsGroupTile, heightOffset := createCollapsibleGroupTile(TileInfo{Title: fmt.Sprintf("GCS Bucket: %s", strings.Split(resourceIds.GcsBucketId, "/")[2])}, gcsMetricsTiles, heightOffset)
```

Comment:
nit: it seems like the same few methods called repeatedly. There may be scope to simplify this or extract into a method to clean up the code